### PR TITLE
[JIT] Add support to register a custom GraphExecutor Implementation

### PR DESCRIPTION
**Pull request by chughtapan**
This change adds support for registering and running custom GraphExecutor Implementations. Currently, extensions can register custom operations and passes, but this only provides a limited interface. This generalizes the way the ProfilingGraphExecutorImpl is used with minimal overheads

### CODE REVIEW

File: torch/csrc/jit/custom_graph_executor_impl.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace torch {
+namespace jit {
+
+using GraphExecutorImplCreator =
+    std::function<GraphExecutorImplBase*(std::shared_ptr<Graph>, bool)>;
+
+static const Symbol kDefaultExecutor =
+    Symbol::fromQualString("executor::default");
+static const Symbol kProfilingExecutor =
```
Comment (apaszke) [inline torch/csrc/jit/custom_graph_executor_impl.h]: Well this puts a `static` symbol into every compliation unit that includes this part, so it's a bit wasteful. Either declare it extern, or just use `fromQualString` in every place that needs this.

File: torch/csrc/jit/custom_graph_executor_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+namespace jit {
+
+namespace {
+
+struct GraphExecutorImplState {
+  Symbol name = kDefaultExecutor;
+  bool impl_cached = false;
+  GraphExecutorImplCreator impl_fn;
+};
+thread_local GraphExecutorImplState executor_state;
```
Comment (apaszke) [inline torch/csrc/jit/custom_graph_executor_impl.cpp]: Hmm I'm not sure if we really want this to be thread local. cc @zdevito

**Review by apaszke — APPROVED** [review]
Looks nice, let's improve the `static` issues though,

File: torch/csrc/jit/custom_graph_executor_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  executor_state.impl_fn = it->second;
+}
+
+RegisterGraphExecutorImpl::RegisterGraphExecutorImpl(
+    Symbol name,
+    GraphExecutorImplCreator creator) {
+  auto& executors = getGraphExecutorImpls();
+  TORCH_CHECK(
+      executors.emplace(name, creator).second,
+      "Cannot register GraphExecutorImpl for ",
```
Comment (apaszke) [inline torch/csrc/jit/custom_graph_executor_impl.cpp]: A better error message would be something along the lines "GraphExecutorImpl for ... has already been registered!"

**Review by apaszke — APPROVED** [review]
This looks good to me, but I'd like @zdevito and @Krovatkin to approve this, since they're working with this code directly.

**Comment by chughtapan** [comment]
@zdevito @Krovatkin Can you please take a look at this?

File: torch/csrc/jit/custom_graph_executor_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  return executors;
+}
+
+} // namespace
+
+Symbol getGraphExecutorName() {
+  return executor_state.name;
+}
+
+void setGraphExecutorName(Symbol name) {
```
Comment (Krovatkin) [inline torch/csrc/jit/custom_graph_executor_impl.cpp]: nitpick: I'm not sure if we need to cache a function pointer. We probably won't have too many executors, so I don't know if we should be worried about optimizing a hash map access. just my 2 cents.

File: torch/csrc/jit/custom_graph_executor_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  executor_state.impl_cached = true;
+  executor_state.impl_fn = it->second;
+}
+
+RegisterGraphExecutorImpl::RegisterGraphExecutorImpl(
+    Symbol name,
+    GraphExecutorImplCreator creator) {
+  auto& executors = getGraphExecutorImpls();
+  TORCH_CHECK(
+      executors.emplace(name, creator).second,
```
Comment (Krovatkin) [inline torch/csrc/jit/custom_graph_executor_impl.cpp]: this probably needs a lock @zdevito @apaszke  ?

File: torch/csrc/jit/custom_graph_executor_impl.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/custom_graph_executor_impl.cpp]: Zach actually insisted making a profiling flag `thread_local`. Although, this isn't exactly the same use pattern 😄

**Review by Krovatkin — CHANGES_REQUESTED** [review]
I'm worried a little bit of a possibility concurrently mutating `executors`

**Review by zdevito — CHANGES_REQUESTED** [review]
Hi sorry for the delay, I only saw this once I was tagged as a reviewer. We are currently in a transition state where we have two executors because we are moving from the original executor to the profiling one. After this transition, the plan of record is to delete all of the virtual dispatch and go back to a single executor. Adding an extension point here is a few lines of code, true, but it is unfortunately a large commitment to support the same interface in the future. Given how much we have changed internals in the past year, I don't think we can provide adequate support for this interface and so I am apprehensive about adding it.

**Comment by jeffreyksmithjr** [comment]
@zdevito I'm unclear on your guidance (because of lack of technical familiarity). Are you saying that we should abandon this whole PR or is your feedback narrower?

**Comment by hidefromkgb** [comment]
Hello everyone.\
Do I get it right that a set of commits closely similar to this PR has already been merged into the main branch?

**Pull request closed without merging**